THU-(38|42): Add preview features settings section

raivieiraadriano92 opened the pull request:
## Description

This PR adds a **Preview Features** section to the Settings screen, allowing users to opt into experimental features.  
- Includes toggles for **Tasks** and **Automations**, disabled by default.  
- If telemetry is disabled, enabling a preview feature triggers a modal requiring users to enable telemetry first.  
- Selecting "Enable telemetry" enables both telemetry and the chosen feature.  
- Navigation items for preview features are feature-gated by toggle state, while the underlying code remains intact.  

---

## Related Issue

Closes
- [THU-38](https://linear.app/mozilla-thunderbolt/issue/THU-38/add-preview-features-settings-section)
- [THU-42](https://linear.app/mozilla-thunderbolt/issue/THU-42/make-tasks-a-preview-feature)

raivieiraadriano92 committed:
feat: add experimental features toggle in preferences settings

- Introduced new analytics events for enabling and disabling experimental features.
- Updated data access layer to fetch experimental features setting.
- Added form schema and handling for experimental features in preferences settings.
- Implemented UI for toggling experimental features with telemetry dependency check.

raivieiraadriano92 committed:
feat: integrate experimental feature wrapper and sidebar visibility toggle

- Added ExperimentalFeatureRouteWrapper to conditionally render the AutomationsPage based on experimental feature settings.
- Updated ChatSidebar to display the Automations link only when experimental features are enabled.
- Enhanced PreferencesSettingsPage with a new section for toggling experimental features, including a telemetry requirement modal.

raivieiraadriano92 committed:
refactor: integrate ExperimentalFeatureWrapper in ChatSidebar

- Removed the use of the boolean setting for experimental features.
- Wrapped the Automations link in ExperimentalFeatureWrapper to manage visibility based on experimental feature settings.

cjroth commented:
This is a good start!

I just realized that the Linear ticket is a bit confusing. There is actually not meant to be a generic toggle for preview features - rather, the Automations feature is meant to be the first feature that has its own toggle. Now that this is built, I think it would make sense to also squeeze https://linear.app/mozilla-thunderbolt/issue/THU-42/make-tasks-a-preview-feature into the same PR. Sorry for the confusion here - I could have done a better job at writing this ticket to be more clear.

A couple other minor UI things:
- We should put this above the "Local Database" and "Privacy" sections IMO
- If the user turns off telemetry and has preview features enabled, I think we should show a modal warning them that the preview features will be disabled as well.
- Now that I see it visually, I think we should skip the yellow warning label - feels a bit much

Not related to this PR but would be nice to fix while we're working on this:
- We should make sure the spacing between each section is the same
- There should be a bit more padding at the bottom of the page so that the last section isn't so tight against the bottom of the viewport

<img width="921" height="623" alt="Screenshot 2025-09-02 at 11 06 46 AM" src="https://github.com/user-attachments/assets/e5724fd3-2fe4-4e55-8258-3aa2a5e0139d" />

raivieiraadriano92 committed:
feat: enhance experimental feature management with new settings

- Updated ExperimentalFeatureWrapper and ExperimentalFeatureRouteWrapper to accept a settingKey prop for better flexibility.
- Refactored preferences settings to include toggles for experimental feature automations and tasks.
- Enhanced analytics events to track enabling and disabling of new experimental features.
- Updated data access layer to fetch and manage new experimental feature settings.

raivieiraadriano92 committed:
feat: improve handling of experimental features in preferences settings

- Added logic to disable experimental features when data collection is turned off.
- Introduced a ref for pending feature toggles to manage state during telemetry modal interactions.
- Updated the save function to handle enabling experimental features based on user actions and current settings.

raivieiraadriano92 committed:
feat: wrap TasksPage in ExperimentalFeatureRouteWrapper and update sidebar

- Integrated ExperimentalFeatureRouteWrapper around TasksPage to manage visibility based on experimental feature settings.
- Updated ChatSidebar to include Tasks link wrapped in ExperimentalFeatureWrapper for conditional rendering based on feature toggles.

raivieiraadriano92 committed:
feat: add telemetry warning modal in preferences settings

- Introduced a modal to warn users that disabling telemetry will also disable all preview features.
- Updated the data collection toggle logic to show the warning modal when telemetry is turned off while experimental features are enabled.
- Removed the previous warning message from the UI for a cleaner presentation.

raivieiraadriano92 committed:
feat: implement preview features management and refactor settings

- Introduced a new configuration for preview features, allowing dynamic management of experimental features like Automations and Tasks.
- Updated the PreferencesSettingsPage to include a dedicated section for toggling preview features with telemetry requirements.
- Refactored the data access layer to fetch and manage preview feature settings more efficiently.
- Enhanced the analytics tracking for enabling and disabling preview features, ensuring accurate event logging.
- Removed legacy experimental feature handling from the preferences settings for cleaner code.

raivieiraadriano92 committed:
fix: adjust padding in settings pages for improved layout

- Increased bottom padding in IntegrationsPage, PreferencesSettingsPage, and ModelsPage to enhance visual spacing and layout consistency.
- Ensured a more balanced appearance across the settings interface.

raivieiraadriano92 committed:
feat: update analytics events for experimental features

- Refined event tracking for enabling and disabling experimental features, specifically for Automations and Tasks.
- Replaced legacy event naming conventions with new identifiers for improved clarity and consistency in analytics logging.

cjroth commented on src/app.tsx:
```diff
@@ -79,8 +81,22 @@ function AppRoutes({ initData }: { initData: InitData }) {
           <Route path="chats/:chatThreadId" element={<ChatDetailPage />} />
-          <Route path="tasks" element={<TasksPage />} />
-          <Route path="automations" element={<AutomationsPage />} />
+          <Route
```
In the ideal world, I would want this to show a placeholder page that says something like "The <feature> feature is not enabled." with a button to go to the settings page and enable it. But I think this is fine for now!

cjroth replied:
Think a little more here... I think the current implementation is unnecessarily complicated. I think we can simplify this by just grabbing the status of the feature in a hook above and then wrapping it in an if statement. I'm not sure if I see the value in a new ExperimentalFeatureRouteWrapper component unless we are going to implement the page I described above

raivieiraadriano92 replied:
done

cjroth commented on src/settings/use-preview-features.ts:
```diff
@@ -0,0 +1,139 @@
+    handleBulkSave,
+    disableAllFeatures,
+    features: PREVIEW_FEATURES,
+  }
```
I think this is a legitimate problem

cjroth commented on src/settings/preview-features-section.tsx:
```diff
@@ -0,0 +1,55 @@
+
+  return (
+    <SectionCard title="Preview Features">
+      <p className="mb-4 text-sm text-muted-foreground">
```
Let's just say `Try out experimental beta features.` I think the rest is implied by the fact that they are beta features.

raivieiraadriano92 replied:
done

cjroth commented on src/settings/preferences.tsx:
```diff
@@ -67,6 +69,9 @@ export default function PreferencesSettingsPage() {
   const [locations, setLocations] = React.useState<LocationData[]>([])
   const [isSearching, setIsSearching] = React.useState(false)
   const [isResetting, setIsResetting] = React.useState(false)
+  const [showTelemetryModal, setShowTelemetryModal] = React.useState(false)
```
I just noticed this - not related to this PR, but we should use `useState` not `React.useState` to be consistent with the rest of the codebase here (this is my fault). Just wanted to say it while it's fresh on my mind.

raivieiraadriano92 replied:
I'll run AI in another branch to refactor the entire project - this way we avoid mix things here

cjroth commented on src/settings/preferences.tsx:
```diff
@@ -67,6 +69,9 @@ export default function PreferencesSettingsPage() {
   const [isResetting, setIsResetting] = React.useState(false)
+  const [showTelemetryModal, setShowTelemetryModal] = React.useState(false)
+  const [showTelemetryWarningModal, setShowTelemetryWarningModal] = React.useState(false)
+  const pendingFeatureToggle = React.useRef<string | null>(null)
```
Not sure that I understand why this one is a ref instead of useState.

raivieiraadriano92 replied:
Usually I use useRef when I need to hold a value which doesn't require a re-render - which is the case here.
However with the useReducer refactor that value was moved to the reducer state for better organization.

cjroth commented on src/settings/preferences.tsx:
```diff
@@ -67,6 +69,9 @@ export default function PreferencesSettingsPage() {
   const [isSearching, setIsSearching] = React.useState(false)
   const [isResetting, setIsResetting] = React.useState(false)
+  const [showTelemetryModal, setShowTelemetryModal] = React.useState(false)
+  const [showTelemetryWarningModal, setShowTelemetryWarningModal] = React.useState(false)
```
I think this page's state would be much cleaner if we switched to useReducer instead of useState given how many useStates we have here. In my experience, prompting the AI to rewrite this with useReducer generally works pretty well on the first try

raivieiraadriano92 replied:
Totally agree - done

cjroth commented on src/settings/preferences.tsx:
```diff
@@ -526,6 +564,60 @@ export default function PreferencesSettingsPage() {
+          <AlertDialogFooter>
+            <AlertDialogCancel>Cancel</AlertDialogCancel>
+            <AlertDialogAction
+              onClick={async () => {
```
This should be defined above the render section like `const handleEnableTelemetry = `

cjroth commented on src/settings/preferences.tsx:
```diff
@@ -526,6 +564,60 @@ export default function PreferencesSettingsPage() {
         </div>
       </SectionCard>
+
+      {/* Telemetry Required Modal */}
```
I wouldn't mind if this were in a separate component file, but don't think it's bad like this either

raivieiraadriano92 replied:
Agreed - it was implemented a new component. It's cleaner now

cjroth commented on src/settings/preferences.tsx:
```diff
@@ -526,6 +564,60 @@ export default function PreferencesSettingsPage() {
+          <AlertDialogFooter>
+            <AlertDialogCancel>Cancel</AlertDialogCancel>
+            <AlertDialogAction
+              onClick={async () => {
```
Same here - should be defined above

raivieiraadriano92 replied:
as above - a new modal component was implemented

cjroth commented on src/settings/use-preview-features.ts:
```diff
@@ -0,0 +1,139 @@
+import React from 'react'
```
Imo, the value that this hook adds doesn't justify the complexity. I feel like the saving logic should be handled in the settings page and doesn't require a hook. And I think the settings should just use `useBooleanSetting` directly.

That said, I DO think we should integrate this with PostHog's "early access features" feature like you described. If we do that, I do think it might make sense to abstract that logic since in that case maybe we are checking multiple locations for the source of truth.

raivieiraadriano92 replied:
I'm working on PostHog's "early access features" implementation in a separate branch - so we keep this PR smaller

cjroth commented on src/settings/preview-features-config.ts:
```diff
@@ -0,0 +1,27 @@
+export enum PreviewFeatureDatabaseKey {
```
Similar to the above comment - I think this just feels like unnecessary abstraction. I don't think we will have enough features that I see a benefit to abstracting this when it could just be hardwired strings like the other settings.

cjroth reviewed (changes requested):
Looks good. Tbh I think I just want it to be a bit simpler overall, but it's the right idea. Nice work!

cjroth commented:
Oh, forgot to add also:

- After I disabled the Tasks feature, the LLM is still calling the "get_tasks" tool. It should not have access to the tasks tools anymore when the feature is disabled.

raivieiraadriano92 committed:
refactor: simplify experimental feature routing in AppRoutes

- Replaced ExperimentalFeatureRouteWrapper with direct conditional rendering for TasksPage and AutomationsPage based on boolean settings.
- Removed the now-unnecessary ExperimentalFeatureRouteWrapper component to streamline the codebase.
- Utilized useBooleanSetting hook for cleaner feature toggling logic.

raivieiraadriano92 committed:
refactor: simplify preview features description in settings

raivieiraadriano92 committed:
refactor: implement useReducer for preferences state management

raivieiraadriano92 committed:
refactor: implemented telemetry modal in its own component

raivieiraadriano92 committed:
fix: update boolean setting for task feature toggle

raivieiraadriano92 committed:
refactor: remove ExperimentalFeatureWrapper and streamline feature toggling

raivieiraadriano92 committed:
refactor: update preferences settings to include experimental feature toggles

raivieiraadriano92 committed:
refactor: consolidate preview features management in PreferencesSettingsPage

raivieiraadriano92 committed:
refactor: enhance event tracking for experimental features in PreferencesSettingsPage

raivieiraadriano92 committed:
refactor: update tool availability logic to incorporate experimental feature toggle

cjroth commented on src/settings/preferences.tsx:
```diff
@@ -230,6 +320,49 @@ export default function PreferencesSettingsPage() {
+  const savePreviewFeaturesMutation = useMutation({
+    mutationFn: async (values: z.infer<typeof previewFeaturesFormSchema>) => {
+      // Save each feature setting
+      await db
```
I feel like we could make a DAL function like setBooleanSetting('experimental_feature_automations', false) or something like that

raivieiraadriano92 replied:
Totally agree - to follow the naming pattern we current have it was implemented `updateBooleanSetting`

cjroth commented on src/lib/dal.ts:
```diff
@@ -151,13 +151,23 @@ export const getPreferencesSettings = async () => {
     preferredName: preferredNameData[0]?.value || '',
     dataCollection: dataCollection[0]?.value === 'false' ? false : true,
+    experimentalFeatureAutomations: experimentalFeatureAutomations[0]?.value === 'false' ? false : true,
+    experimentalFeatureTasks: experimentalFeatureTasks[0]?.value === 'false' ? false : true,
```
I think I agree with this that they should default to false. We could do that as part of the next PR though

cjroth reviewed (approved):
Looks great!! Love this.

raivieiraadriano92 committed:
refactor: introduce updateBooleanSetting utility for streamlined feature toggling

raivieiraadriano92 committed:
refactor: replace direct database mutation with updateBooleanSetting

raivieiraadriano92 committed:
refactor: await updateBooleanSetting in preferences to ensure proper async handling

raivieiraadriano92 commented on src/settings/preferences.tsx:
```diff
@@ -230,6 +314,30 @@ export default function PreferencesSettingsPage() {
+      // Save each feature setting
+      updateBooleanSetting('experimental_feature_automations', values.experimentalFeatureAutomations)
+      updateBooleanSetting('experimental_feature_tasks', values.experimentalFeatureTasks)
+    },
```
done
